File: src/aoc/day01.clj
(ns aoc.day01
  (:require
   [clojure.string :as str]))

(def input
  (->> (slurp "data/input01.txt")
       str/split-lines
       (map read-string)))

(defn part-1 []
  (reduce + input))

(defn find-seen [{:keys [freq seen] :as acc} change]
  (let [freq (+ freq change)]
    (if (seen freq)
      (reduced freq)
      (assoc acc :freq freq :seen (conj seen freq)))))

(defn part-2 []
  (reduce find-seen {:freq 0 :seen #{0}} (cycle input)))

(comment
  (part-1)
  (part-2))

File: elixir/day01.exs
defmodule Day01 do
  def parse_input(filename) do
    {:ok, input} = File.read(filename)

    input
    |> String.split("\n", trim: true)
    |> Enum.map(&String.to_integer/1)
  end

  def find_repeating(change, {freq, seen}) do
    freq = freq + change

    if MapSet.member?(seen, freq) do
      {:halt, freq}
    else
      {:cont, {freq, MapSet.put(seen, freq)}}
    end
  end

  def part1(input) do
    Enum.sum(input)
  end

  def part2(input) do
    Enum.reduce_while(
      Stream.cycle(input),
      {0, MapSet.new()},
      &find_repeating/2
    )
  end
end

input = Day01.parse_input("../data/input01.txt")

input |> Day01.part1() |> IO.puts()
input |> Day01.part2() |> IO.puts()
